File: 50-powx-n/50-powx-n.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n < 0:
            return 1 / self.myPow(x, -n)
        if n == 0:
            return 1
        if n == 1:
            return x
        if n == 2:
            return x * x
        ans = self.myPow(x, n >> 1)
        ans = ans * ans
        if n % 2 == 1: ans *= x
        return ans

File: 50-powx-n/50-powx-n.cpp
class Solution {
public:
  double myPow(double x, int n) {
    return n >= 0 ? myPowImp(x, n) : 1.0 / myPowImp(x, n);
  }
private:
  double myPowImp(double x, int n) {
    if (n == 0) return 1;
    return myPowImp(x * x, n / 2) * (n % 2 ? x : 1);
  }
};